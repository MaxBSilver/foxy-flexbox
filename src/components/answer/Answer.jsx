import React, { Component } from "react";


export default class Answer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      jcVal: "",
      acVal: "",
      btnColor: '#404b69'
    };
  }

  handleChange = e => {
    const { value, name } = e.target;
    this.setState({ [name]: value }, () => {
      this.props.updateUserAnswer(this.state);
      if (
        this.state.jcVal === this.props.justifyContent &&
        this.state.acVal === this.props.alignContent
      ) {
        this.setState({btnColor: '#65a565'})
      }
    })
  };


  handleSubmit = e => {
    e.preventDefault();
    if (
      this.state.jcVal === this.props.justifyContent &&
      this.state.acVal === this.props.alignContent
    ) {
      this.props.determineQuestion();
      this.props.incrementRound();
      this.setState({ jcVal: "", acVal: "", btnColor: '#404b69'});
    }
  };
  

  render() {
    const btnColor = this.state.btnColor;
    return (
          <section className="answer--container">
            <section className="code-snippet--container">
              <form onSubmit={this.handleSubmit}>
                <div className="form--container">
                  <p className="code-snippet--title"> #garden{" {"} </p>
                  <div className="user-form">
                    <p className="code-snippet--body">display : flex;</p>
                    <p className="code-snippet--body">
                      flex-direction: {this.props.flexDirection};
                    </p>
                    <div className="input-one">
                      <p className="code-prompt"> justify-content : </p>
                      <input
                        type="text"
                        className="search-input"
                        name="jcVal"
                        value={this.state.jcVal}
                        onChange={this.handleChange}
                        placeholder=""
                      />
                    </div>
                    <div className="input-two">
                      <p className="code-prompt"> align-content : </p>
                      <input
                        type="text"
                        className="search-input"
                        name="acVal"
                        value={this.state.acVal}
                        onChange={this.handleChange}
                        placeholder=""
                      />
                    </div>
                  </div>
                  <p className="code-snippet--title">{"}"}</p>
                </div>
                <div className="submit--container">
                  <input type="submit" className="search-btn" value="Next" style={{backgroundColor : btnColor }}/>
                </div>
              </form>
            </section>
          </section>
    );
  }
}

