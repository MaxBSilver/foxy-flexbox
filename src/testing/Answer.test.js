import React from "react";
import Answer from "../components/answer/Answer";
import { shallow } from "enzyme";

const mockDetermineQuestion = jest.fn();
const mockIncrementRound = jest.fn();
const mockUpdateUserAnswer = jest.fn();
const mockedEvent = { target: { value: "value", name: "name" } };
const mockQuestionA = {
  id: 1,
  difficulty: "easy",
  display: "",
  flexDirection: "row",
  prompt: "justify-content",
  justifyContent: "",
  alignContent: "",
  flexWrap: "",
  btnColor: "#65a565"
};
const mockQuestionB = {
  id: 22,
  difficulty: "easy",
  display: "random",
  flexDirection: "row",
  prompt: "justify-content",
  justifyContent: "random",
  alignContent: "random",
  btnColor: "#65a565",

  flexWrap: "random"
};
const mockQuestionC = {
  id: 23,
  difficulty: "easy",
  display: "",
  flexDirection: "row",
  prompt: "justify-content",
  justifyContent: "",
  alignContent: "",
  btnColor: "#65a565",

  flexWrap: ""
};

describe("Answer", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Answer
        {...mockQuestionA}
        determineQuestion={mockDetermineQuestion}
        incrementRound={mockIncrementRound}
        updateUserAnswer={mockUpdateUserAnswer}
      />
    );
  });
  afterEach(() => {
    wrapper = shallow(
      <Answer
        {...mockQuestionA}
        determineQuestion={mockDetermineQuestion}
        incrementRound={mockIncrementRound}
      />
    );
  });
  it("should match the snapshot with all data passed in", () => {
    expect(wrapper).toMatchSnapshot();
  });
  describe("handleSubmit", () => {
    it("should not invoke determineQuestion and incrementRound when the ID is under 21 and the state does not match the question", () => {
      wrapper = shallow(<Answer {...mockQuestionB} />);
      wrapper.find("form").simulate("submit", { preventDefault: () => {} });
      expect(mockDetermineQuestion).not.toHaveBeenCalled();
      expect(mockIncrementRound).not.toHaveBeenCalled();
    });
    it("should invoke determineQuestion and increment round when the ID is under 21 and state matches question", () => {
      wrapper.find("form").simulate("submit", { preventDefault: () => {} });
      expect(mockDetermineQuestion).toHaveBeenCalled();
      expect(mockIncrementRound).toHaveBeenCalled();
    });
    it("should invoke determineQuestion and increment round when the ID is over 21 and state matches question", () => {
      wrapper = shallow(
        <Answer
          {...mockQuestionC}
          determineQuestion={mockDetermineQuestion}
          incrementRound={mockIncrementRound}
        />
      );
      wrapper.find("form").simulate("submit", { preventDefault: () => {} });
      expect(mockDetermineQuestion).toHaveBeenCalled();
      expect(mockIncrementRound).toHaveBeenCalled();
    });
  });
  describe("render HardSnippet", () => {
    it("should not render HardSnippet unless round is over 20", () => {
      expect(wrapper.find("HardSnippet").exists()).toBeFalsy();
      wrapper = shallow(<Answer {...mockQuestionB} />);
      expect(wrapper.find("HardSnippet").exists()).toBeTruthy();
    });
  });
  describe("handleChange", () => {
    it("should invoke updateAnswer when called", () => {
      wrapper = shallow(
        <Answer
          {...mockQuestionA}
          determineQuestion={mockDetermineQuestion}
          incrementRound={mockIncrementRound}
          updateUserAnswer={mockUpdateUserAnswer}
        />
      );
      wrapper.instance().handleChange(mockedEvent);
      expect(mockUpdateUserAnswer).toHaveBeenCalled();
    });
    it("should invoke updateAnswer when called", () => {
      wrapper = shallow(
        <Answer
          {...mockQuestionC}
          determineQuestion={mockDetermineQuestion}
          incrementRound={mockIncrementRound}
          updateUserAnswer={mockUpdateUserAnswer}
        />
      );
      wrapper.instance().handleChange(mockedEvent);
      expect(mockUpdateUserAnswer).toHaveBeenCalled();
      expect(wrapper.state("btnColor")).toEqual("#65a565");
    });
    it("should invoke updateAnswer when called", () => {
      wrapper = shallow(
        <Answer
          {...mockQuestionA}
          determineQuestion={mockDetermineQuestion}
          incrementRound={mockIncrementRound}
          updateUserAnswer={mockUpdateUserAnswer}
        />
      );
      wrapper.instance().handleChange(mockedEvent);
      expect(mockUpdateUserAnswer).toHaveBeenCalled();
      expect(wrapper.state("btnColor")).toEqual("#65a565");
    });
  });
});
