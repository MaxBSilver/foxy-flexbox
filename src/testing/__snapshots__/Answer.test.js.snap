// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer should match the snapshot with all data passed in 1`] = `
<section
  className="answer--container"
>
  <section
    className="code-snippet--container"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form--container"
      >
        <p
          className="code-snippet--title"
        >
           #garden
           {
           
        </p>
        <div
          className="user-form"
        >
          <p
            className="code-snippet--body"
          >
            display : flex;
          </p>
          <p
            className="code-snippet--body"
          >
            flex-direction: 
            row
            ;
          </p>
          <EasySnippet
            acVal=""
            btnColor="$darkBlueBackground"
            fwVal=""
            handleChange={[Function]}
            jcVal=""
          />
          <p
            className="code-snippet--title"
          >
            }
          </p>
        </div>
        <div
          className="submit--container"
        >
          <input
            className="search-btn"
            style={
              Object {
                "backgroundColor": "$darkBlueBackground",
              }
            }
            type="submit"
            value="Next"
          />
        </div>
      </div>
    </form>
  </section>
</section>
`;
